src/mlpack/methods/ne/neat.hpp

```diff
@@ -0,0 +1,972 @@
+      if (mlpack::math::Random() < p) {
+        ////printf("mutate 3\n");
+        MutateAddLink(genome, aMutateAddLinkProb, aMutateAddRecurrentLinkProb, aMutateAddLoopLinkProb);
+        ////printf("mutate 4\n");
+      }
+      --p;
+    }
+
+    // Mutate neuron
+    ////printf("mutate 5\n");
+    p = aMutateAddNeuronProb;
+    ////printf("mutate 6\n");
+    while (p > 0) {
+      if (mlpack::math::Random() < p) {
+        ////printf("mutate 7\n");
+        MutateAddNeuron(genome, aMutateAddNeuronProb);
+        ////printf("mutate 8\n");
+      }
+      --p;
+    }
+
+    // Mutate enabled node to disabled.
+    p = aMutateEnabledProb;
+    while (p > 0) {
+      if (mlpack::math::Random() < p) {
+       ////printf("mutate 9\n");
+        MutateEnableDisable(genome, true, aMutateEnabledProb);
+        ////printf("mutate 10\n");
+      }
+      --p;
+    }
+
+    // Mutate disabled node to enabled.
+    p = aMutateDisabledProb;
+    while (p > 0) {
+      if (mlpack::math::Random() < p) {
+        ////printf("mutate 11\n");
+        MutateEnableDisable(genome, false, aMutateDisabledProb);
+        ////printf("mutate 12\n");
+      }
+      --p;
+    }
+  }
+
+  // Breed child for a species.
+  // NOTICE: can have different ways to breed a child.
+  bool BreedChild(Species& species, Genome& childGenome, double crossoverProb) {
+    ////printf("breed 1\n");
+    double p = mlpack::math::Random();
+    ssize_t speciesSize = species.aGenomes.size();
+    ////printf("species size is %d \n", speciesSize);
+    ////printf("breed 2\n");
+    if (speciesSize == 0)
+      return false;
+
+    if (p < crossoverProb) {
+      ssize_t idx1 = mlpack::math::RandInt(0, speciesSize);
+      ssize_t idx2 = mlpack::math::RandInt(0, speciesSize);
+      ////printf("idx1 is %d\n", idx1);
+      ////printf("idx2 is %d\n", idx2);
+      ////printf("breed 3\n");
+      Crossover(species.aGenomes[idx1], species.aGenomes[idx2], childGenome);
+      ////printf("breed 4\n");
+    } else {
+      ssize_t idx = mlpack::math::RandInt(0, speciesSize);
+      ////printf("breed 5\n");
+      childGenome = species.aGenomes[idx];
+      ////printf("breed 6\n");
+    }
+    ////printf("breed 7\n");
+    Mutate(childGenome);
+    ////printf("breed 8\n");
+    return true;
+  }
+
+
+  // Initialize population.
+  void InitPopulation() {
+    aPopulation = Population(aSeedGenome, aPopulationSize);
+  }
+
+  // Reproduce next generation of population.
+  void Reproduce() {
+    // keep best genome.
+    Genome lastBestGenome = aPopulation.BestGenome();
+
+    // Remove stale species.
+    if (aPopulation.aSpecies.size() > 10) {
+      RemoveStaleSpecies(aPopulation);
+    }
+    ////printf("hehe 0\n"); // DEBUG
+    // Remove weak genomes in each species.
+    CullSpecies(aPopulation, aCullSpeciesPercentage);
+    ////printf("hehe 1\n"); // DEBUG
+
+    // Remove weak species.
+    if (aPopulation.aSpecies.size() > 10) {
```

Comment:
As for the number of stale species, we should also provide a parameter for the number of weak species.